### PR TITLE
Fix mmcv-dataparallel

### PR DESCRIPTION
**Pull request by ZwwWayne**
Fix https://github.com/open-mmlab/mmdetection/issues/1501
Fix https://github.com/open-mmlab/mmdetection/issues/3048

A RetinaNet is tested with three modes: multi-GPU distributed testing, single-GPU testing with MMDataParallel, and CPU testing with MMDataParallel. The model achieves similar AP under these three modes.

### CODE REVIEW

File: mmcv/parallel/data_parallel.py [hunk 1]
```diff
@@ -1,18 +1,90 @@
...
+
+    MMDataParallel has two main differences with PyTorch DataParallel:
+
+    - It supports a custom type :class:`DataContainer` which allows more
+      flexible control of input data during both GPU and CPU inference.
+    - It implement two more APIs ``train_step()`` and ``val_step()``.
+
+    Args:
+        module (:class:`nn.Module`): Module to be encapsulated.
+        device_ids (list[int]): Device IDS of modules to be scattered to.
```
Comment (xvjiarui) [inline mmcv/parallel/data_parallel.py:23]: We may add docs about CPU cases as well.

Comment (xvjiarui) [reply mmcv/parallel/data_parallel.py:23]: We may also check if the user input negative number, only `[-1]` is allowed?

File: mmcv/parallel/_functions.py [hunk 2]
```diff
@@ -19,8 +19,11 @@ def scatter(input, devices, streams=None):
         output = input.contiguous()
         # TODO: copy to a pinned buffer first (if copying from CPU)
         stream = streams[0] if output.numel() > 0 else None
-        with torch.cuda.device(devices[0]), torch.cuda.stream(stream):
-            output = output.cuda(devices[0], non_blocking=True)
+        if devices != [-1]:
+            with torch.cuda.device(devices[0]), torch.cuda.stream(stream):
+                output = output.cuda(devices[0], non_blocking=True)
+        # This is a workaround
+        output = output.unsqueeze(0)
```
Comment (xvjiarui) [inline mmcv/parallel/_functions.py:26]: This will affect original behavior.

File: mmcv/parallel/data_parallel.py [hunk 3]
```diff
@@ -1,18 +1,93 @@
...
+                    'module must have its parameters and buffers '
+                    f'on device {self.src_device_obj} (device_ids[0]) but '
+                    f'found one of them on device: {t.device}')
+
+        inputs, kwargs = self.scatter(inputs, kwargs, self.device_ids)
+        if len(self.device_ids) == 1:
+            return self.module(*inputs[0], **kwargs[0])
+        replicas = self.replicate(self.module, self.device_ids[:len(inputs)])
+        outputs = self.parallel_apply(replicas, inputs, kwargs)
+        return self.gather(outputs, self.output_device)
```
Comment (hellock) [inline mmcv/parallel/data_parallel.py:81]:
```python
if not self.device_ids:
    # our own codes
else:
    super().forward(xxx)
```

File: mmcv/parallel/data_parallel.py [hunk 4]
```diff
@@ -1,18 +1,93 @@
...
+    """
+
+    def __init__(self, module, device_ids=None, output_device=None, dim=0):
+        super(DataParallel, self).__init__()
+
+        if not torch.cuda.is_available():
+            self.module = module
+            self.device_ids = []
+            # we add this line from the original DP to enable CPU inference.
+            self.dim = dim
```
Comment (hellock) [inline mmcv/parallel/data_parallel.py:36]: Only this line is needed.

File: mmcv/parallel/scatter_gather.py [hunk 5]
```diff
@@ -15,7 +15,11 @@ def scatter(inputs, target_gpus, dim=0):
 
     def scatter_map(obj):
         if isinstance(obj, torch.Tensor):
-            return OrigScatter.apply(target_gpus, None, dim, obj)
+            if target_gpus != [-1]:
+                return OrigScatter.apply(target_gpus, None, dim, obj)
+            else:
+                # for CPU inference we use self-implemented scatter
+                return Scatter.forward(target_gpus, obj.data)
```
Comment (hellock) [inline mmcv/parallel/scatter_gather.py:22]: Why `obj.data`?

**Pull request merged**